code/json.sls

```diff
@@ -30,172 +30,218 @@
+              (begin (display "," p)
+                     (if pretty?
+                         (display "\n" p)
+                         (display " " p)))
+              (set! need-comma #t))
+          (let* ((entry (car vec))
+                 (k (car entry))
+                 (v (cdr entry)))
+            (when pretty? (display-level p))
+            (display "\"" p)
+            (cond
+             ((symbol? k)
+              (display (string-replace (symbol->string k) #\- #\_) p))
+             ((string? k)
+              (display k p))
+             (else (error "Invalid JSON table key in json-write" k)))
+            (display "\": " p)
+            (write-any v p)))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n")
+          (display-level p))
+        (display "}" p))
 
       (define (write-array a p)
-	(display "[" p)
-	(let ((need-comma #f))
-	  (for-each (lambda (v)
-		      (if need-comma
-			  (display ", " p)
-			  (set! need-comma #t))
-		      (write-any v p))
-		    a))
-	(display "]" p))
+        (display "[" p)
+        (when pretty?
+          (display "\n" p)
+          (set! indent-level (+ indent-level 1)))
+        (let ((need-comma #f))
+          (for-each (lambda (v)
+                      (if need-comma
+                          (if pretty?
+                              (display ",\n" p)
+                              (display "," p))
+                          (set! need-comma #t))
+                      (when pretty? (display-level p))
+                      (write-any v p))
+                    a))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n")
+          (display-level p))
+        (display "]" p))
 
       (define (write-any x p)
-	(cond
-	  ((or (string? x)
-	       (number? x)) (write x p))
-	  ((boolean? x) (display (if x "true" "false") p))
-	  ((symbol? x) (write (if (eq? x 'null) 'null (symbol->string x))
-			      p)) ;; for convenience
-	  ((null? x) (display "null" p))
-	  ((and (list? x)
-		(pair? (car x))
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
-
-      (lambda (x . maybe-port)
-	(write-any x (if (pair? maybe-port) (car maybe-port) (current-output-port))))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
+          (write (if (eq? x 'null)
+                     'null
+                     (string-replace (symbol->string x) #\- #\_))
+                 p)) ;; for convenience
+         ((null? x) (display "null" p))
+         ((and (list? x)
+               (pair? (car x))
+               (not (pair? (caar x))))
+          (write-ht x p))
+         ((list? x) (write-array x p))
+         (else (error "Invalid JSON object in json-write" x))))
+
+      (lambda (msg x . maybe-port)
```

Comment:
I think this is a sub-par interface for the writer.  @jitwit I know you suggested just keeping both functions, but it will literally drive me completely insane to break `DRY` principles that badly.

I am going to try re-implementing this interface lambda with `lambda-case` in a way that handles everything gracefully.  I think it will work pretty well.  I'll have some more work on that up later today :smiling_imp:

Reply:
Haha, fair enough. Sounds good

Reply:
I'm about to have dinner with my mother for her birthday, but I'll get back to this later tonight.

Reply:
I still haven't done this yet.  I went ahead and wrote the preprocessor extensions though, so at least that is out of the way.

Reply:
Started conversation about the new interface in a comment below